precompiles/batch/src/lib.rs

```diff
@@ -0,0 +1,337 @@
+                        handle.record_log_costs(&[&log])?;
+                        log.record(handle)?;
+
+                        match mode {
+                            Mode::BatchAll => {
+                                return Err(PrecompileFailure::Error {
+                                    exit_status: ExitError::OutOfGas,
+                                })
+                            }
+                            Mode::BatchSomeUntilFailure => {
+                                return Ok(succeed(EvmDataWriter::new().write(true).build()))
+                            }
+                            Mode::BatchSome => continue,
+                        }
+                    }
+                    limit
+                }
+            };
+
+            let (reason, output) = handle.call(
+                address,
+                transfer,
+                call_data,
+                Some(forwarded_gas),
+                false,
+                &sub_context,
+            );
+
+            // Logs
+            // We reserved enough gas so this should not OOG.
+            match reason {
+                ExitReason::Revert(_) | ExitReason::Error(_) => {
+                    let log = log_subcall_failed(handle.code_address(), i);
+                    handle.record_log_costs(&[&log])?;
+                    log.record(handle)?
+                }
+                ExitReason::Succeed(_) => {
+                    let log = log_subcall_succeeded(handle.code_address(), i);
+                    handle.record_log_costs(&[&log])?;
+                    log.record(handle)?
+                }
+                _ => (),
+            }
+
+            // How to proceed
+            match (mode, reason) {
+                // _: Fatal is always fatal
+                (_, ExitReason::Fatal(exit_status)) => {
+                    return Err(PrecompileFailure::Fatal { exit_status })
+                }
+
+                // BatchAll : Reverts and errors are immediatly forwarded.
+                (Mode::BatchAll, ExitReason::Revert(exit_status)) => {
+                    return Err(PrecompileFailure::Revert {
+                        exit_status,
+                        output,
+                    })
+                }
+                (Mode::BatchAll, ExitReason::Error(exit_status)) => {
+                    return Err(PrecompileFailure::Error { exit_status })
+                }
+
+                // BatchSomeUntilFailure : Reverts and errors prevent subsequent subcalls to
+                // be executed but the precompile still succeed.
+                (Mode::BatchSomeUntilFailure, ExitReason::Revert(_) | ExitReason::Error(_)) => {
+                    return Ok(succeed(EvmDataWriter::new().write(true).build()))
+                }
+
+                // Success or ignored revert/error.
+                (_, _) => (),
+            }
+        }
+
+        Ok(succeed(EvmDataWriter::new().write(true).build()))
```

Comment:
God function! 🙂

I understand it's code reuse, but try checking the UT coverage, just in case.
There are a lot of branches to account for.

Reply:
It has been designed carefully, we only allow it to pass for the function like `BatchSome` or `BatchSomeUntilFailure` and not for `BatchAll`.

Reply:
I'm sure it was but still, it's an incredibly complex function - is it fully covered by UTs?
You can use tarpaulin tool to check this: `cargo tarpaulin -o html`
(it's not a perfect tool but gives good enough picture)

We are just reusing this, we hadn't really written it from scratch. It's easy to overlook something when just reading/reusing the code.

Reply:
I ran tarpaulin and I think it is a good score.
```precompiles/batch/src/lib.rs: 127/143```

I checked all the tests again, we have enough tests imo, MB has done good testing.

Reply:
I wouldn't know to say whether that's a good score or not.

But checking coverage in %, and even better, taking a look at the output file, gives a clearer picture if some important function/branch was missed.